### ISSUE TITLE
Open in browser not working

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#12209]
### Describe your problem!

We have two users on support mail with this problem:
If I want to open a cache in my browser I get an error toast "No browser found". It was working with 2021.10.08 and stopped working with the new version.

Both users have in common, that they use some "non-standard" browser (1st user: VIA browser, 2nd user: Samsung Browser) and do not have Chrome installed.

Chrome Web view is switched off for both users.

Both users confirm, that the browser is opening if they use the function to open a geocaching users profile in their browser from c:geo.

Its not related to Android 12, both users are on Android 11.

I failed to reproduce this issue.

### How to reproduce?

- Open cache details
- Select Menu -> Open in Browser

### Actual result after these steps?

Error toast "No browser found" (or similar)

### Expected result after these steps?

Browser should open respectively (if more than one is present) browser selection should open

### Reproducible

No

### c:geo Version

2021.11.21

### System information

```text
1st user:
c:geo version: 2021.11.16-RC

Device:
-------
- Device type: SM-A405FN (a40eea, samsung)
- Available processors: 8
- Android version: 11
- Android build: RP1A.200720.012.A405FNXXS4CUH2
- Screen resolution: 1080x2110px (360x703dp)
- Pixel density: 3.0
- System font scale: 1.0 / used scale: 1.0
- Sailfish OS detected: false
- Google Play services: disabled - 21.42.18 (150408-410302452)

Sensor and location:
-------
- Low power mode: inactive
- Compass capabilities: yes
- Rotation vector sensor: present
- Orientation sensor: present
- Magnetometer & Accelerometer sensor: present
- Direction sensor used: rotation vector

Program settings:
-------
- Settings: v8, Count:165
- Set language: sk_SK (system default)
- System date format: d. M. y
- Time zone: GMT+01:00
- Debug mode active: no
- Last backup: 18. 10., 23:20
- Routing mode: Car
- Live map mode: true
- OSM multi-threading: true / threads: 4
- Map: Google: Satelit
  - Id: cgeo.geocaching.maps.google.v2.GoogleMapProvider$GoogleSatelliteSource
  - Atts: none
  - Theme: paws_4/paws_4.xml

Filters:
-------
- Hide waypoints: original visited
- LIVE: Typ kešky:11 položiek,Stav:3 položky ([:inconclusive=false:advanced=true]AND(type:WHERIGO:LETTERBOX:MULTI:EARTH:VIRTUAL:TRADITIONAL:USER_DEFINED:CITO:MYSTERY:WEBCAM:EVENT;difficulty_terrain:d=-:d=-:t=-:t=-;status:owned_no:found_no:solved_mystery_yes))
- OFFLINE: Žiadne ([:inconclusive=false:advanced=false]AND(type;difficulty_terrain:d=-:d=-:t=-:t=-;status))

Services:
-------
- Geocaching sites enabled:
   geocaching.com: Logged in (Prihlásenie úspešné) / PREMIUM
   Geocaching.com Adventure Lab
- Geocaching.com date format: dd.MM.yyyy
- Routing: internal / BRouter installed: false
- Installed c:geo plugins: contacts

Permissions & paths:
-------
- Fine location permission: granted
- Write external storage permission: granted
- System internal c:geo dir: /data/user/0/cgeo.geocaching (23,0 GB free) v2 internal isDir(10 entries)
- Legacy User storage c:geo dir: /storage/emulated/0/cgeo (23,0 GB free) v2 external non-removable isDir(7 entries)
- Geocache data: /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData (23,0 GB free) v2 external non-removable isDir(1453 entries)
- Internal theme sync (is turned ON): /data/user/0/cgeo.geocaching/MapThemeData (23,0 GB free) v2 internal isDir(4 entries)
- Public Folders: #11
  - BASE: /cgeo (User-Defined)[/cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo, Av:true, files:>=97, dirs:>=8, totalFileSize:>=95,7 MB, free space: 23,0 GB, files on device: 1691648)
  - OFFLINE_MAPS: Karta SD/osm (User-Defined)[Karta SD/osm[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/012F-1341%3Aosm::]] (Uri: content://com.android.externalstorage.documents/tree/012F-1341%3Aosm/document/012F-1341%3Aosm, Av:true, files:>=1030, dirs:>=5, totalFileSize:>=3,2 GB, free space: 7,2 GB, files on device: 0)
  - OFFLINE_MAP_THEMES: Karta SD/osm (User-Defined)[Karta SD/osm[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/012F-1341%3Aosm::]] (Uri: content://com.android.externalstorage.documents/tree/012F-1341%3Aosm/document/012F-1341%3Aosm, Av:true, files:>=1030, dirs:>=5, totalFileSize:>=3,2 GB, free space: 7,2 GB, files on device: 0)
  - LOGFILES: /cgeo/logfiles (Default)[/cgeo/logfiles[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/logfiles]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Flogfiles, Av:true, files:4, dirs:0, totalFileSize:56,2 MB, free space: 23,0 GB, files on device: 1691648)
  - GPX: /cgeo/gpx (Default)[/cgeo/gpx[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/gpx]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fgpx, Av:true, files:2, dirs:0, totalFileSize:1,7 MB, free space: 23,0 GB, files on device: 1691648)
  - BACKUP: /cgeo/backup (Default)[/cgeo/backup[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/backup]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fbackup, Av:true, files:2, dirs:1, totalFileSize:37,5 MB, free space: 23,0 GB, files on device: 1691648)
  - FIELD_NOTES: /cgeo/field-notes (Default)[/cgeo/field-notes[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/field-notes]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Ffield-notes, Av:true, files:89, dirs:0, totalFileSize:255,9 KB, free space: 23,0 GB, files on device: 1691648)
  - SPOILER_IMAGES: /cgeo/GeocachePhotos (Default)[/cgeo/GeocachePhotos[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/GeocachePhotos]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2FGeocachePhotos, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 23,0 GB, files on device: 1691648)
  - ROUTING_BASE: /cgeo/routing (Default)[/cgeo/routing[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting, Av:true, files:9, dirs:1, totalFileSize:137,0 MB, free space: 23,0 GB, files on device: 1691648)
  - ROUTING_TILES: /cgeo/routing/segments4 (Default)[/cgeo/routing/segments4[PERSISTABLE_FOLDER(ROUTING_BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing/segments4]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting%2Fsegments4, Av:true, files:2, dirs:0, totalFileSize:136,9 MB, free space: 23,0 GB, files on device: 1691648)
  - TEST_FOLDER: [Legacy]/data/user/0/cgeo.geocaching/files/unittest (Default)[/data/user/0/cgeo.geocaching/files/unittest[FILE#1:p-file:///data/user/0/cgeo.geocaching/files::/unittest]] (Uri: file:///data/user/0/cgeo.geocaching/files/unittest, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 23,0 GB, files on device: -1)
- Map render theme path: paws_4/paws_4.xml
- PersistedDocumentUris: #1
- TRACK: null
- Persisted Uri Permissions: #2
  - content://com.android.externalstorage.documents/tree/primary%3Acgeo (19. 3., 18:19):RW
  - content://com.android.externalstorage.documents/tree/012F-1341%3Aosm (19. 3., 18:19):RW
- Database: /data/user/0/cgeo.geocaching/databases/data (v98, Size:42,8 MB) on system internal storage

--- End of system information ---



2nd user:


## System information

c:geo version: 2021.11.21

Device:- Device type: SM-A515F (a51nseea, samsung)
- Available processors: 8
- Android version: 11
- Android build: RP1A.200720.012.A515FXXS5EUI4
- Screen resolution: 1080x2186px (411x832dp)
- Pixel density: 2.3625
- System font scale: 0.9 / used scale: 0.9
- Sailfish OS detected: false
- Google Play services: enabled - 21.42.18 (150400-410302452)

Sensor and location:- Low power mode: inactive
- Compass capabilities: yes
- Rotation vector sensor: present
- Orientation sensor: present
- Magnetometer & Accelerometer sensor: present
- Direction sensor used: rotation vector

Program settings:- Settings: v8, Count:163
- Set language: de_DE (system default)
- System date format: dd.MM.yy
- Time zone: GMT+01:00
- Debug mode active: no
- Last backup: 10. Nov., 20:11
- Routing mode: Walk
- Live map mode: true
- OSM multi-threading: true / threads: 4
- Map: Germany (Offline)
- Id: cgeo.geocaching.maps.mapsforge.MapsforgeMapProvider$OfflineMapSource:9016-4EF8:cgeo/germany.map
- Atts: Map data (c) OpenStreetMap contributors
- Theme: none

Filters:- Hide waypoints: -
- LIVE: Status:3 Elemente ([:inconclusive=false:advanced=false]AND(type;difficulty_terrain:d=-:d=-:t=-:t=-;status:owned_no:found_no:exclude_archived))
- OFFLINE: Keine ([:inconclusive=false:advanced=false]AND(type;difficulty_terrain:d=-:d=-:t=-:t=-;status))

Services:- Geocaching sites enabled:
geocaching.com: Logged in (Anmeldung OK) / PREMIUM
Geocaching.com Adventure Lab
- Geocaching.com date format: dd/MM/yyyy
- Routing: external / BRouter installed: false
- Installed c:geo plugins: none

Permissions & paths:- Fine location permission: granted
- Write external storage permission: granted
- System internal c:geo dir: /data/user/0/cgeo.geocaching (97,1 GB free) v2 internal isDir(8 entries)
- Legacy User storage c:geo dir: /storage/emulated/0/cgeo (97,0 GB free) v2 external non-removable isDir(7 entries)
- Geocache data: /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData (97,0 GB free) v2 external non-removable isDir(6 entries)
- Internal theme sync (is turned ON): /data/user/0/cgeo.geocaching/MapThemeData (97,1 GB free) v2 internal isDir(0 entries)
- Public Folders: #11
- BASE: /cgeo (User-Defined)[/cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo, Av:true, files:>=7, dirs:>=11, totalFileSize:>=158,1 MB, free space: 97,0 GB, files on device: 13278715)
- OFFLINE_MAPS: SD-Karte/cgeo (User-Defined)[SD-Karte/cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/9016-4EF8%3Acgeo::]] (Uri: content://com.android.externalstorage.documents/tree/9016-4EF8%3Acgeo/document/9016-4EF8%3Acgeo, Av:true, files:6, dirs:1, totalFileSize:2,3 GB, free space: 27,2 GB, files on device: 0)
- OFFLINE_MAP_THEMES: SD-Karte/cgeo/_themes (Default)[SD-Karte/cgeo/_themes[PERSISTABLE_FOLDER(OFFLINE_MAPS)#1:p-content://com.android.externalstorage.documents/tree/9016-4EF8%3Acgeo::/_themes]] (Uri: content://com.android.externalstorage.documents/tree/9016-4EF8%3Acgeo/document/9016-4EF8%3Acgeo%2F_themes, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 27,2 GB, files on device: 0)
- LOGFILES: /cgeo/logfiles (Default)[/cgeo/logfiles[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/logfiles]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Flogfiles, Av:true, files:10, dirs:0, totalFileSize:2,0 MB, free space: 97,0 GB, files on device: 13278715)
- GPX: 3364-6431/cgeo (User-Defined)[3364-6431/cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/3364-6431%3Acgeo::]] (Uri: content://com.android.externalstorage.documents/tree/3364-6431%3Acgeo/document/3364-6431%3Acgeo, Av:false, files:0, dirs:0, totalFileSize:0 B, free space: -1 B, files on device: -1)
- BACKUP: /cgeo/backup (Default)[/cgeo/backup[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/backup]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fbackup, Av:true, files:7, dirs:4, totalFileSize:158,1 MB, free space: 97,0 GB, files on device: 13278715)
- FIELD_NOTES: /cgeo/field-notes (Default)[/cgeo/field-notes[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/field-notes]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Ffield-notes, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 97,0 GB, files on device: 13278715)
- SPOILER_IMAGES: /cgeo/GeocachePhotos (Default)[/cgeo/GeocachePhotos[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/GeocachePhotos]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2FGeocachePhotos, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 97,0 GB, files on device: 13278715)
- ROUTING_BASE: /cgeo/routing (Default)[/cgeo/routing[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting, Av:true, files:7, dirs:1, totalFileSize:81,2 KB, free space: 97,0 GB, files on device: 13278715)
- ROUTING_TILES: /cgeo/routing/segments4 (Default)[/cgeo/routing/segments4[PERSISTABLE_FOLDER(ROUTING_BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing/segments4]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting%2Fsegments4, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 97,0 GB, files on device: 13278715)
- TEST_FOLDER: [Legacy]/data/user/0/cgeo.geocaching/files/unittest (Default)[/data/user/0/cgeo.geocaching/files/unittest[FILE#1:p-file:///data/user/0/cgeo.geocaching/files::/unittest]] (Uri: file:///data/user/0/cgeo.geocaching/files/unittest, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 97,1 GB, files on device: -1)
- Map render theme path:
- PersistedDocumentUris: #1
- TRACK: null
- Persisted Uri Permissions: #3
- content://com.android.externalstorage.documents/tree/3364-6431%3Acgeo (29. März, 05:31):RW
- content://com.android.externalstorage.documents/tree/9016-4EF8%3Acgeo (15. Nov., 16:55):RW
- content://com.android.externalstorage.documents/tree/primary%3Acgeo (29. März, 05:30):RW
- Database: /data/user/0/cgeo.geocaching/databases/data (v98, Size:65,7 MB) on system internal storage

--- End of system information ---
```

### Additional Information

Ticket 473058
Ticket 476719

**Comment by Lineflyer** [comment]
User with VIA browser provided a video, which shows the problem:
 https://youtu.be/sIa6-KnsIOk

**Comment by eddiemuc** [comment]
**Analysis**

The difference in code between opening a cache listing in browser and opening a user profile in browser boils down to how the following method is called:

https://github.com/cgeo/cgeo/blob/b7c0aece98fb9bac87777888eb8b52308c2a80fe/main/src/cgeo/geocaching/utils/ShareUtils.java#L149

* "Open (cache) in browser" calls this method with `forceIntentChooser` set to `true`
* "Open profile in browser" calls this method with `forceIntentChooser` set to `false`

In the video, when the user tries to open cache in browser, he gets the error toast "No browser found". This uniquely points to our custom browser tab (implemented in `StartWebviewActivity`). It means that our method `ProcessUtils.getInstalledBrowsers(this)` returns zero results. This in turn means that calling the following lines return zero browsers:

```
//the following code returns zero results on the affected system
final Intent browserIntent = new Intent("android.intent.action.VIEW", Uri.parse("https://example.com"));
return context.getPackageManager().queryIntentActivities(browserIntent, PackageManager.MATCH_DEFAULT_ONLY);
```

In conclusion, the affected system seems not to expose callable activities to handle VIEW intends to the application.

On the other hand, simply calling the intent action VIEW without further hand-written logic is honoured by the system (which is proven by "open profile in browser" to work):

```
//the following code works on the affected system
final Intent viewIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
...
context.startActivity(viewIntent);
```

**Comment by eddiemuc** [comment]
Conclusion

I noticed a rather complex "intent chooser" logic inside `ShareUtils.openUrl`. Why is this necessary? Wouldn't it be feasible to simply open our custom tab if `Settings.getUseCustomTabs() && ProcessUtils.isChromeLaunchable()` (as is already implemented) and otherwise simply call startActivity for ACTION_VIEW (without any further "chooser" logic)? 

If this chooser logic is really necessary, my second suggestion would be to introduce a switch to turn if off - for devices not supplying necessary activity information (as the ones in this issue). Yes, that would be a new setting unfortunately.

**Comment by eddiemuc** [comment]
**Another idea**

While thinking about this I had another idea: maybe the constant Intent.VIEW simply has another value than "android.intent.action.VIEW" on the affected systems? This would certainly explain why no browsers are returned.

I will provide a small PR to change this hard-coded value against usage of the constant.

**Comment by Lineflyer** [comment]
AFAIR the complex logic was invented/created to avoid getting into an endless loop (or just open another instance of c:geo) if the user assigned c:geo as default to open "https://coord.info/..." links. This would also explain why its only associated for opening cache pages.

Additional note:
Both users claim it worked (with intent chooser) before the current release. So its likely that some change in our code caused the changed behavior (however it could also be a coincidence).

**Comment by eddiemuc** [comment]
@Lineflyer understood. 

Then let's dig deeper:

If `Settings.getUseCustomTabs()` is set then the logic defaults to the c:geo custom tab (no chooser or default browser is opened regardless of flag `forceIntentChooser`) - but only if Chrome launchable is available.

However, if `Settings.getUseCustomTabs()` is **not** set and  `forceIntentChooser= true`, the c:geo custom tab is **always** included as the first chooser option. `ProcessUtils.isChromeLaunchable()` is not checked in this case. Is this correct behaviour?

(In case of this issue the custom tab clearly does not work no the users system - hence the error message "No Browser found")

@fm-sys it seems you implemented this back in the days. Is there any reason for this behaviour?

**Comment by eddiemuc** [comment]
@Lineflyer I have a seemingly weird request to the users affected by this problem: could they maybe try the following:
* type the following URL (exactly like this) into some app e.g. a chat app a la whatsapp or SMS app: https://example.com 
* Assumingly this app recognizes this as a link: try to open the link from the app
* => question is: does their system auto-recognizes the URL and does their system automatically opens the URL in a browser?

(I am asking this weird request because c:geo internally uses this exact hard-coded URL to search for system providers which can open it via action VIEW. Those providers are then assumed to be "available browsers", and currently on the user's system it seems that no such browser is found)

**Comment by fm-sys** [comment]
I think I have an idea what the problem could be:

Android 11 has introduced a new package visibility handling... Therefore `ProcessUtils.getInstalledBrowsers` doesn't work anymore. It was quite some back and forth to get this finally working, so let's hope we can find a new equivalent solution. 

https://developer.android.com/training/package-visibility

**Comment by fm-sys** [comment]
On stack overflow, they suggest a stupidly simple solution: The only thing which needs to be done is to add `browserIntent.addCategory(Intent.CATEGORY_BROWSABLE)` to the intent.

https://stackoverflow.com/questions/22957823/android-open-url-in-external-browser-causes-an-infinite-loop


But probably we have to remove this line in our manifest intent filter to make it work (which does exist since the very first initial commit):
```
<category android:name="android.intent.category.BROWSABLE" />
```

---

Another solution which might be worth a try: It is possible to exclude a specific app from being shown in the chooser since API 24...
```
val components = arrayOf(ComponentName(applicationContext, YourActivity::class.java))
if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) 
   startActivity(Intent.createChooser(intent, null).putExtra(Intent.EXTRA_EXCLUDE_COMPONENTS,components))
```
https://newbedev.com/how-to-exclude-a-specific-application-from-action-send-intent

**Comment by Lineflyer** [comment]
IMHO I don't think there is real need for change, but some strange device or situation specific issue.

Today I tried to simulate the problem using an Android 11 emulator without Google services (thus without Chrome) and installed tag market_20211121 on it.

On first try of "Open in browser" I got a "No browser found" toast, which is correct in that case.
Then I installed the via browser on this emulator and now I get this, which looks perfectly fine:

![image](https://user-images.githubusercontent.com/949669/144512849-42f7d2eb-bb23-409a-940f-011d97074dda.png)

**Comment by Lineflyer** [comment]
@eddiemuc 
> I have a seemingly weird request 

Sounds reasonable. I forwarded this question to both users.

> I will provide a small PR to change this hard-coded value against usage of the constant.

Such kind of potential device specific issue is, what I would also assume right now.
I cannot judge however if this might help...could be worth a try for next bugfix release to check it out if no other idea comes up.

@fm-sys 
> Android 11 has introduced a new package visibility handling...

I don't think this is the problem here, as we would have tons of support mail in such a generic case.
AFAIU we don't ask for a specific package (where it is indeed a problem we solved already e.g. for navigation apps where we need to check the install status) in this case but ask the system to tell us the browsers. I think Google was clever enough not to hide browsers from other apps.

Anyway I would wait with larger changes until we have
a) Feedback to the test
b) Maybe applied the small trial as suggested by @eddiemuc first and get feedback once the bugfix release is out.

**Comment by eddiemuc** [comment]
@Lineflyer 
> Then I installed the via browser on this emulator and now I get this, which looks perfectly fine

That's a good idea to try. I have a follow up question: if you click on the selection "c:geo open website" (after you installed the via browser), does this work?

**Comment by Lineflyer** [comment]
> That's a good idea to try. I have a follow up question: if you click on the selection "c:geo open website" (after you installed the via browser), does this work?

That only opens the cache (again) in c:geo.

**Comment by Lineflyer** [comment]
> @Lineflyer I have a seemingly weird request to the users affected by this problem: could they maybe try the following:
> 
> * type the following URL (exactly like this) into some app e.g. a chat app a la whatsapp or SMS app: https://example.com
> * Assumingly this app recognizes this as a link: try to open the link from the app
> * => question is: does their system auto-recognizes the URL and does their system automatically opens the URL in a browser?

Reply received from both users:
- One user tried in Telegram and it worked as expected (default browser is opened)
- The other user tried in some "chat messenger" (does not name it) and it also works as expected (default browser is opened)

**Comment by Lineflyer** [comment]
I will build a "preRC" version containing the fix @eddiemuc applied and provide it to them to get early feedback if anything changed to the better.

**Comment by eddiemuc** [comment]
> I will build a "preRC" version containing the fix @eddiemuc applied and provide it to them to get early feedback if anything changed to the better.
> 

I wouldn't get my hope up too much. It's a very long shot...

**Comment by eddiemuc** [comment]
I still have no better idea than to introduce yet another setting for these users, to enforce direct default action (not using a chooser)

**Comment by Lineflyer** [comment]
The user with "VIA browser" just replied:
The new version has the same problem.

Additionally he sent some videos, where he did further tests by installing Chrome and testing again.
From that I could clearly see, that the VIA browser is not recognized as installed browser by the app chooser:
As soon as Chrom is available the app chooser shows up when users tries "Open in browser" but only contains Chrome and c:geo itself but not VIA.

I asked the user for permission to share the video here, so that you can see the test he did.

Its really strange, that VIA is not recognized, its works in my emulator tests.

**Comment by Lineflyer** [comment]
I was finally able to reproduce it:
- Have an emulator without Google apps (or disable Chrome and all other browsers)
- Install VIA and set it as default
- Now it still works as expected
- Now to go the c:geo app settings and select "Open as default" -> Always open links in c:geo
- Now the error toast "No browser found" appears when selecting "Open in browser"

It seems the exact reason why the app chooser solution was implemented (to bock looping back to c:geo when a link supported by c:geo shall be opened) turns out to have a negative effect in this scenario. 

Will test what happens if I install a second or other browser in the same scenario as it was never a problem for me to let c:geo be the default to open supported links.

**Comment by Lineflyer** [comment]
Tested now on my real device:
- Had Chrome already installed
- Installed Opera
- Installed Firefox
- Installed Via
- Selected c:geo as default to open supported links
- Open cache - "open in browser"

Result:
The app selection dialog is shown but only contains Chrome and c:geo itself.

So the bug seems to drill down to the fact, that we do not "see" any of the other browsers although they are installed.
So if Chrome is not present and defaults are set to c:geo it will result in "No browser found".

Could it indeed be an issue of packet visibility then?

**Comment by moving-bits** [comment]
> Could it indeed be an issue of packet visibility then?

I've created a PR to test this - added Firefox, Samsung and Via browsers to our manifest to make their packages visible to c:geo. Can you cross-check whether this helps?

**Comment by Lineflyer** [comment]
As reported in that PR the test was positive:
The package visibility limitation makes our app chooser fail.

But before we start just adding a lot of browser packages into our manifest:
Isn't there any option we could use as last resort to query the package name of the default browser of the system via any method?

**Comment by moving-bits** [comment]
Updated #12240 to implement the generic browser query in contrast to listing the extra browsers.
